### PR TITLE
Integrating /proposal route and improved proposal editor page

### LINKED ISSUE #473: Proposal functionality inside Donut - Frontend.
When a developer or a certain user wants to suggest/propose an idea to an open-source
organization it mostly ends up being communicated through IRC messages or other
informal ways. In most cases their suggestions/proposals go unanswered or ignored
among hundreds of messages in the IRC resulting in the original user getting frustrated.

The issue here is that there is no clear cut procedure/pathway that a user can follow in
order to propose a project to the upper echelons of an organization

This project at its core focuses on eliminating that issue by providing the users with a
built-in platform. With this platform the users can create, edit, publish and share the
proposals directly with the organization admins and obtain crucial feedback

The frontend designs for this project could be found [here](https://www.figma.com/file/iZvb7rLYRp3Di3wILVQ7V9/CodeUino-Donut?node-id=1664%3A3).

### PR DESCRIPTION
**Pull request by AuraOfDivinity**
Partially fixes #473 

Implementation of the demonstration done on 11/06/20

Improvements to the proposal editor and proposal dashboard pages.
Integrates the usage of backend /proposal routes.

Kindly review @vaibhavdaren @devesh-verma

### CODE REVIEW

**Comment by vaibhavdaren** [comment]
@AuraOfDivinity pls fix comments

File: src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js [hunk 1]
```diff
@@ -2,12 +2,143 @@ import React, { Component } from "react";
...
 import "./DashboardContent.scss";
 import userIcon2 from "../../../../images/userIcon2.jpg";
+import { Link } from "react-router-dom";
 
 class DashboardContent extends Component {
   constructor(props) {
     super(props);
-    this.state = {};
+    this.state = {
+      token:
```
Comment (Rupeshiya) [inline src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js:11]: @AuraOfDivinity Please remove this token from here. Already token is saved in the localStorage. Fetch from there after login

File: src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js [hunk 2]
```diff
@@ -2,12 +2,143 @@ import React, { Component } from "react";
...
+      displayItems: [],
+    };
   }
+
+  componentDidMount() {
+    this.getProposalsByUserId();
+    this.getAllProposals();
+  }
+
+  getAllProposals() {
```
Comment (Rupeshiya) [inline src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js:24]: Make these as actions in actions folder, and dispatch from there to update your state.

File: src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js [hunk 3]
```diff
@@ -2,12 +2,143 @@ import React, { Component } from "react";
...
+      .then((resData) => {
+        console.log(resData);
+        this.setState({
+          proposals: resData.proposals,
+          displayItems: resData.proposals,
+        });
+      });
+  }
+
+  getProposalsByUserId() {
```
Comment (Rupeshiya) [inline src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js:47]: same here

Comment (Rupeshiya) [reply src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js:47]: Also, no need to write http://localhost/:5000 coz the request is already proxied to that port in package.json. Please check once

File: src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js [hunk 4]
```diff
@@ -25,150 +157,118 @@ class DashboardContent extends Component {
...
-                  </Button>
-                </div>
-              </div>
             </div>
           </div>
         </div>
+
+        <div className="proposal-container">
+          <div className="proposals">
+            {this.state.displayItems.map((proposalItem, index) => {
```
Comment (Rupeshiya) [inline src/user/proposals/AdminProposalDashboard/DashboardContent/DashboardContent.js:207]: This will throw error when this.state.displayItems will be empty [], try once. Handle it

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js [hunk 5]
```diff
@@ -1,32 +1,78 @@
...
     super(props);
-    this.state = {};
+    this.state = {
+      commentContent: "",
+      commentItems: this.props.commentItems,
+    };
   }
 
+  handleComment = (text) => {
+    fetch("http://localhost:5000/proposal/comment", {
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js:16]: Same here, make it as seperate action, also use axios as it's already used in most of the part of integration, It's stable as well.

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js [hunk 6]
```diff
@@ -1,24 +1,104 @@
...
+    );
   }
+
+  getData = () => {
+    console.log("getDtata called");
+    fetch("http://localhost:5000/proposal/" + this.state.proposalId, {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: this.state.token,
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js:57]: No need to configure fetch or axios every time, just configure once for all the requests to the server

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js [hunk 7]
```diff
@@ -96,104 +174,118 @@ class DiscussionContent extends Component {
...
+                  <Modal onClose={this.toggleModal}>
+                    <Carousel views={this.state.images} />
+                  </Modal>
+                ) : null}
+              </ModalGateway>
             </div>
           </div>
           <div className="comments">
-            <DiscussionComments commentItems={this.state.comments} />
+            <DiscussionComments
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js:276]: Better to use destructuring

File: src/user/proposals/ProposalEditor/EditorContent/DropZone.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+  const {
+    getRootProps,
+    getInputProps,
+    isDragActive,
+    isDragAccept,
+    isDragReject,
+  } = useDropzone({ onDrop, accept: "image/*" });
+
+  const style = useMemo(
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalEditor/EditorContent/DropZone.js:73]: Great!! 👍

File: src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js [hunk 9]
```diff
@@ -2,12 +2,119 @@ import React, { Component } from "react";
...
 import userIcon2 from "../../../../images/userIcon2.jpg";
+import { Link } from "react-router-dom";
 
 class DashboardContent extends Component {
   constructor(props) {
     super(props);
-    this.state = {};
+    this.state = {
+      token:
+        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZWRjYjE4ZjIxNWRhNzRjMThkM2YyNzQiLCJpYXQiOjE1OTE1MjE2Nzl9.q3g5Ah_rtjPIrH7z183fVmUBTv_A4OjEoL673zeG250",
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js:12]: remove

File: src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js [hunk 10]
```diff
@@ -2,12 +2,119 @@ import React, { Component } from "react";
...
+      proposals: [],
+      displayItems: [],
+    };
   }
+
+  componentDidMount() {
+    this.getProposalsByUserId();
+  }
+
+  getProposalsByUserId() {
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js:23]: actions

File: src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js [hunk 11]
```diff
@@ -2,12 +2,119 @@ import React, { Component } from "react";
...
+import { Link } from "react-router-dom";
 
 class DashboardContent extends Component {
   constructor(props) {
     super(props);
-    this.state = {};
+    this.state = {
+      token:
+        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZWRjYjE4ZjIxNWRhNzRjMThkM2YyNzQiLCJpYXQiOjE1OTE1MjE2Nzl9.q3g5Ah_rtjPIrH7z183fVmUBTv_A4OjEoL673zeG250",
+      userId: "5edcb18f215da74c18d3f274",
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardContent/DashboardContent.js:13]: remove hardcoded data

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+import userIcon2 from "../../../../../images/userIcon2.jpg";
+import socket from "../Utils/socket";
+import notifications from "../../../../dashboard/notifications/notifications";
+
+class Notifications extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      socket: socket,
+      token:
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:13]: remove

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZWRjYjE4ZjIxNWRhNzRjMThkM2YyNzQiLCJpYXQiOjE1OTE1MjE2Nzl9.q3g5Ah_rtjPIrH7z183fVmUBTv_A4OjEoL673zeG250",
+      userId: "5edcb18f215da74c18d3f274",
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:15]: remove

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    this.state.socket.on("proposal deleted", (data) => {
+      data.newNotification = true;
+      data.createdAt = new Date();
+      this.setState({
+        notifications: [...this.state.notifications, data],
+      });
+    });
+  }
+
+  fetchNotifications = () => {
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:40]: remove from here

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            ],
+          },
+          () => {
+            this.fetchUserNotifications();
+          }
+        );
+      });
+  };
+
+  fetchUserNotifications = () => {
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:66]: same

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+      this.setState({
+        notifications: [...this.state.notifications, data],
+      });
+    });
+
+    this.state.socket.on("proposal deleted", (data) => {
+      data.newNotification = true;
+      data.createdAt = new Date();
+      this.setState({
+        notifications: [...this.state.notifications, data],
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:35]: new notifications should be on the top, please check if is this breaking my code functionality

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+      return new Date(b.createdAt) - new Date(a.createdAt);
+    });
+
+    this.setState({
+      Notifications: notifications,
+    });
+  };
+
+  render() {
+    const notifications = this.state.notifications;
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:109]: better to use destructing for better flexibility

**Comment by vaibhavdaren** [comment]
fix conflicts

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js (hunk 13 above)
Comment (vaibhavdaren) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Notifications/Notifications.js:15]: please follow coding practices from other prs.

File: src/user/proposals/ProposalEditor/EditorContent/EditorContent.js [hunk 18]
```diff
@@ -428,4 +450,12 @@ class EditorContent extends Component {
   }
 }
 
-export default withRouter(EditorContent);
+const mapStateToProps = (state) => ({
+  createdProposal: state.createdProposal,
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalEditor/EditorContent/EditorContent.js:454]: Here is the issue @AuraOfDivinity 
change state.createdProposal to state.proposal.createdProposal  and
change state.fetchedProposal to state.proposal.fetchedProposal

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js (hunk 7 above)
Comment (AuraOfDivinity) [reply src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js:231]: Do we get advantages by destructuring here @Rupeshiya ?

Comment (Rupeshiya) [reply src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionContent.js:231]: Not for now, but think if you want to use other state property in the component, then what will you prefer everytime writing this.state.property or only property after destructuring??
@AuraOfDivinity

File: .vscode/launch.json [hunk 19]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (Rupeshiya) [inline .vscode/launch.json:1]: @AuraOfDivinity What is this for?

File: src/actions/types.js [hunk 20]
```diff
@@ -1,12 +1,14 @@
...
-export const  RESPONSE_MSG = "RESPONSE_MSG";
+export const GET_CURRENT_USER = "GET_CURRENT_USER";
+export const SET_CURRENT_USER = "SET_CURRENT_USER";
+export const RESPONSE_MSG = "RESPONSE_MSG";
 export const GET_ERRORS = "GET_ERRORS";
 export const SET_ERROR = "SET_ERROR";
 export const SET_STATUS = "SET_STATUS";
 export const GET_PLATFORM_NOTIFICATIONS = "GET_PLATFORM_NOTIFICATIONS";
 export const GET_USER_NOTIFICATIONS = "GET_USER_NOTIFICATIONS";
-export const GET_ALL_NOTIFICATIONS = "GET_ALL_NOTIFICATIONS";
```
Comment (Rupeshiya) [inline src/actions/types.js:9]: @AuraOfDivinity It's not used anywhere GET_ALL_NOTIFICATIONS??

File: src/user/dashboard/insights/components/memberInfo.scss [hunk 21]
```diff
@@ -136,27 +136,27 @@
...
 .list-group {
-  width: 14vw;
+  //  width: 14vw;
 }
 .popover-body {
   width: 9vw;
   text-align: center;
   .overlay_list_btn {
     cursor: pointer;
-     background: transparent;
```
Comment (Rupeshiya) [inline src/user/dashboard/insights/components/memberInfo.scss:146]: @AuraOfDivinity Also please take care of this, I think your extension is making some unnecessary changes

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js [hunk 22]
```diff
@@ -1,14 +1,45 @@
...
+    this.state = {
+      commentContent: "",
+      commentItems: this.props.commentItems,
+      imageModalOpen: false,
+    };
   }
 
+  handleComment = (text) => {
+    const commentData = {
+      userId: this.props.userId,
```
Comment (Rupeshiya) [inline src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js:20]: destructure, please 😄

File: src/user/proposals/UserProposalDashboard/DashboardRightPanel/Utils/socket.js [hunk 23]
```diff
@@ -0,0 +1,3 @@
+import io from "socket.io-client";
```
Comment (Rupeshiya) [inline src/user/proposals/UserProposalDashboard/DashboardRightPanel/Utils/socket.js:1]: Why you have created another instance of socet.io??
As it's already there

Comment (Rupeshiya) [reply src/user/proposals/UserProposalDashboard/DashboardRightPanel/Utils/socket.js:1]: @AuraOfDivinity

File: .vscode/launch.json (hunk 19 above)
Comment (AuraOfDivinity) [reply .vscode/launch.json:1]: @Rupeshiya It was for debugging purposes. My Bad I shouldn't have committed it.

File: src/user/dashboard/insights/components/memberInfo.scss (hunk 21 above)
Comment (AuraOfDivinity) [reply src/user/dashboard/insights/components/memberInfo.scss:146]: @Rupeshiya I'm not exactly sure why it's like this. Basically nothing is getting changed in this file...

File: src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js (hunk 22 above)
Comment (Rupeshiya) [reply src/user/proposals/ProposalDiscussion/DiscussionContent/DiscussionComments/DiscussionComments.js:20]: use const { userId } = this.props
Then use userId: userId

**Comment by AuraOfDivinity** [comment]
@Rupeshiya @vaibhavdaren @devesh-verma I have made all the requested changes. Please review.

File: src/actions/notificationAction.js [hunk 24]
```diff
@@ -1,41 +1,63 @@
...
 export const getAllNotifications = () => async (dispatch) => {
   try {
-    const res = await axios.get('/notification/org/all')
-    dispatch(setRequestStatus(false))
+    const res = await axios.get("/notification/org/all");
+    dispatch(setRequestStatus(false));
     if (res.status === 200) {
-      dispatch(setRequestStatus(true))
-      console.log('Whole platform notification ', res.data.notifications)
+      dispatch(setRequestStatus(true));
```
Comment (vaibhavdaren) [inline src/actions/notificationAction.js:16]: are these formatting changes or code changes by linter did you make these?

Comment (Rupeshiya) [reply src/actions/notificationAction.js:16]: @vaibhavdaren These are the due to the extensions he is using or due to his vs code settings.
I have already mentioned that

File: src/user/dashboard/insights/components/memberInfo.scss (hunk 21 above)
Comment (vaibhavdaren) [reply src/user/dashboard/insights/components/memberInfo.scss:146]: @AuraOfDivinity can you please fix this

**Comment by jaskiratsingh2000** [comment]
Checks failing
